### PR TITLE
feat(appeals): investigate ability to bulk download

### PR DESCRIPTION
**Pull request by BenSurgisonPINS**
Bulk download spike

WEB:
- Add a link that downloads a zipped file containing all uploaded documents retaining the folder structure
- Manually tested within browser

## Issue ticket number and link

[A2-1467 Spike - Investigate ability to bulk download from Azure storage](https://pins-ds.atlassian.net/browse/A2-1467)

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [x] I have provided a description of the ticket
- [ ] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: appeals/web/src/server/app/components/file-downloader.component.js [hunk 1]
```diff
@@ -198,3 +197,106 @@ export const getDocumentDownloadByVersion = async ({ apiClient, params, session
...
+					document.latestDocumentVersion;
+				return {
+					fullName: `${folder.path}/${document.name}`,
+					blobStorageContainer,
+					blobStoragePath,
+					documentURI
+				};
+			});
+		});
+
```
Comment (fabfab-kc) [inline appeals/web/src/server/app/components/file-downloader.component.js:267]: Here there should be a check for an empty `bulkFileInfo` (i.e. no files available).

File: appeals/web/src/server/app/components/file-downloader.component.js [hunk 2]
```diff
@@ -198,3 +197,106 @@ export const getDocumentDownloadByVersion = async ({ apiClient, params, session
...
+	archive.pipe(response);
+
+	const blobStorageClient = await createBlobStorageClient(session);
+
+	if (!bulkFileInfo || !bulkFileInfo.length) {
+		return response.status(404);
+	}
+
+	const blobStreams = await Promise.all(
```
Comment (fabfab-kc) [inline appeals/web/src/server/app/components/file-downloader.component.js:285]: There's a small chance a document record may not have the corresponding blob (especially on dev machines): if that happens, the stream gets locked... maybe append to the archive only if the blob stream is valid?

File: appeals/web/src/server/app/components/file-downloader.component.js [hunk 3]
```diff
@@ -198,3 +197,106 @@ export const getDocumentDownloadByVersion = async ({ apiClient, params, session
...
+ * Build a zipped file and Download
+ *
+ * @param {{apiClient: import('got').Got, params: {caseId: string, filename?: string}, session: SessionWithAuth}} request
+ * @param {Response} response
+ * @returns {Promise<Response>}
+ */
+export const getBulkDocumentDownload = async ({ apiClient, params, session }, response) => {
+	const { filename: requestedFilename, caseId } = params;
+
+	// console.log(JSON.stringify({ getBulkDocumentDownload1: { caseId, requestedFilename } }, null, 2));
```
Comment (fabfab-kc) [inline appeals/web/src/server/app/components/file-downloader.component.js:249]: remove

**Review by fabfab-kc — COMMENTED** [review]
This looks great @BenSurgisonPINS - added a few comments on the best way to move it forward and take it out of draft

File: appeals/web/src/server/app/components/file-downloader.component.js (hunk 1 above)
Comment (fabfab-kc) [reply appeals/web/src/server/app/components/file-downloader.component.js:267]: We should also have the ability to filter-out folders... e.g. internal and costs folders should probably be excluded (those will be the left part of the `folder.path` which usually indicates the area and the document type)

File: appeals/web/src/server/appeals/appeal-details/appeal-details.mapper.js [hunk 4]
```diff
@@ -43,6 +52,7 @@ export async function appealDetailsPage(
 		...buildNotificationBanners(session, 'appealDetails', appealDetails.appealId),
 		...(await generateStatusTags(mappedData, appealDetails, request)),
 		generateCaseSummary(mappedData),
+		allDocumentsDownloadLink,
```
[the remark this answers is not in the document]
Comment (fabfab-kc) [reply appeals/web/src/server/appeals/appeal-details/appeal-details.mapper.js:55]: We would also to consider generally permissions, i.e. who can download the docs, or maybe even more in detail, i.e. which folder depending on permissions...

File: appeals/web/src/server/app/components/file-downloader.component.js (hunk 2 above)
Comment (BenSurgisonPINS) [reply appeals/web/src/server/app/components/file-downloader.component.js:278]: I'm filtering out the empty blobStreams when appending to the archive to fix this issue.

Comment (BenSurgisonPINS) [reply appeals/web/src/server/app/components/file-downloader.component.js:278]: If there are any missing files, they will be reported within the missing-files.json file at the root of the zipped file

**Pull request closed without merging**